### ISSUE TITLE
StatePaintR

### ISSUE THREAD

**Issue by Simon-Coetzee** [Bioconductor/Contributions#348]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/Simon-Coetzee/StatePaintR

Confirm the following by editing each check box to '[x]'

- [x] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [x] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [x] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [x] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [x] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [x] The 'devel' branch for new packages and features.
- [x] The stable 'release' branch, made available every six
      months, for bug fixes.
- [x] _Bioconductor_ version control using [Subversion][5]
  (optionally [via GitHub][6]).

For help with submitting your package, please subscribe and post questions
to the [bioc-devel][4] mailing list.

[1]: https://bioconductor.org/developers/package-guidelines/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/source-control/
[6]: http://bioconductor.org/developers/how-to/git-mirrors/

**Comment by lawremi** [comment]
Some initial comments. `PaintStates()` should accept an R object containing the manifest, not the file path (user can read that in an initial step). That would be more flexible, as the user could programmatically define a manifest. The object returned by `PaintStates()` should contain the decision matrix used, to simplify the code and avoid inconsistencies. The vignette processes a lot of files using low-level routines. It should probably use existing Bioconductor functions (e.g., for parsing BED files use rtracklayer), or the package should provide the functions, at least for the common sources. The package can automatically load packages that users require to use the API by listing them in Depends. Probably too many libraries are loaded at the beginning of the vignette.

**Comment by Simon-Coetzee** [comment]
Hi @lawremi ,
1. `PaintStates()` is able to take a `data.frame()` as a manifest, and the documentation has been updated to reflect this.
2. You are right, for simplicity and for avoiding inconsistency, the output of `PaintStates()` now includes the manifest and the `decisionMatrix`, and the documentation has been updated to reflect this.
3. I have drastically cut down and simplified the vignette, hopefully this the correct direction.

Thanks for the suggestions.

**Comment by lawremi** [comment]
It seems like there is a lot of good stuff in the vignette, so good that it could go into the package, and potentially rely more on existing infrastructure. I wasn't suggesting to simply delete sections from the vignette.

**Comment by Simon-Coetzee** [comment]
I implemented a new function called `PRG()` which wraps up a lot of the precision recall gain analysis and documented it in the package. I also reintroduced the full comparison of enhancer calling.

The old plots looking for enrichment along genes and active and inactive enhancers from ensembl are still gone though. It relied on lots of processing outside of R, so I'd rather not include it in the vignette. It can still be found on the StateHub website, that the package interacts with.

**Comment by Simon-Coetzee** [comment]
hey @lawremi 

What do you think of the current incarnation of the vignette? I'm happy to receive more suggestions.

**Comment by lawremi** [comment]
It seems there is still some code that could be wrapped up into high-level functions in the package, particularly stuff for parsing different data sources. But I'll let @nturaga assess things further.

**Comment by nturaga** [comment]
StatePaintR Review
==================

## Build

### April 25th 2017

R-release CMD build StatePaintR: Ok

R-release CMD INSTALL StatePaintR: Ok


## DESCRIPTION

* Good DESCRIPTION section; clear abstract

* Good reuse of Bioconductor packages; biocViews.

## NAMESPACE

* ok. 

## Vignette

* Downloading vignette data takes about a minute, i'm guessing its not a small dataset. Running vignette also takes some time. Consider using [BiocFileCache][] to reduce the need to re-download resources.

* In terms of functionality in the vignette, it seems like a lot of the code chunks are creating `GRanges()` from files, and extra lines of code are introduced because the files do not have standardized headers or formats. Why not create the files with appropriate headers or in formats like BED, and use `GRanges(read_tsv())` or `rtracklayer::import()` to simplify these code chunks? This would help emphasize use of the functions you wrote, rather than standard R data processing commands.

* Some sections of the vignette could use more description (section 4.4 onwards).

[BiocFileCache]: https://bioconductor.org/packages/BiocFileCache

## man

* All man pages should have executable examples, especially as there are no unit tests.
* Could use some more info on some of the functions (e.g., decisionMatrix)

## R

* Reconsider class names so that they reflect available functionality, e.g., `decisionMatrix` does not have matrix-like (`dim()`, `[,i,j`) functionality.

* Variable naming is not_consistent (`get.decision.matrix` vs `ExportStatePaintR` vs `doNotScore`). Please refer to bioconductor or R code style document and refractor the code to follow those rules. 

* Some of the code, e.g., `PaintStates` is very hard to parse. Some suggestions include:

    - use [formatR][] to create consistent indentation and line length
    - separate nested code into indepedent functions. This can help make the code more readable, and can also help in testing.
    - remove commented out code
    - clearly separate 'tests' at the start of the function from code
      evaluation, e.g,. statepaintr.R:56 simply test (and fail if not)
      for whether the object is a decisionMatrix.
    - consider `vapply()` rather than `sapply()` for more robust iteration.
    - 'copy-and-append' paradigms like `rbind()` (statepaintr.R:384)
      or `c()` are very inefficient when there are many elements in a
      loop or nested loop. Consider [pre-allocate and fill][] or other
      strategies, especially looking for opportunities to vectorize
      rather than iterate.
    - Use `is()` to test class type for S4 classes.
    - Use `tryCatch()` rather than `try()` for handling error conditions.
    - Avoid direct slot access `@` even for internal code; write
      accessor functions (e.g., plain-old functions that are not
      exported, `.abstraction.layer <- function(x)
      x@abstraction.layer`.
    - Document internal functions using roxygen-style notation, but
      line prefix `# ` rather than `#'`. This is particularly
      important for functions like `.create.crossing.points()` in
      prg.R that are central to the package.

[formatR]: https://cran.r-project.org/package=formatR
[pre-allocate and fill]: http://bioconductor.org/developers/how-to/efficient-code/

**Comment by nturaga** [comment]
Hi @Simon-Coetzee ,

Just wanted to follow up this package. Any updates?

**Comment by nturaga** [comment]
Hi @Simon-Coetzee 

If you are making changes in the package, please reply by Friday (06/09/2017). Otherwise, we'll close the issue, since there has been no feedback.

Best,

**Comment by Simon-Coetzee** [comment]
@nturaga Is it possible to reactivate this issue, I have more time to work on issues with the package.

**Comment by nturaga** [comment]
Hi @Simon-Coetzee 

We have reopened the issue as you requested. Please let me know when you have fixed issues and bumped the version. I will review your package again.

**Comment by nturaga** [comment]
Hi @Simon-Coetzee ,

Just checking if you are actively developing this package, or if you are ready for review. Your package needs a version bump to go on our build machines.

**Comment by mtmorgan** [comment]
OK, back to closing this issue; easy to re-open when appropriate.
